library/asn1write.c

```diff
@@ -287,26 +287,73 @@ int mbedtls_asn1_write_ia5_string( unsigned char **p, unsigned char *start,
     return( (int) len );
 }
 
+int mbedtls_asn1_write_named_bitstring( unsigned char **p,
```

Comment:
Non-blocking suggestion:

Functionally, this looks fine. My only objection is that this function could be written in a more compact way, thereby saving some bytes of code. I appreciate that most of the library isn't optimized in that regard, but I think we should try to raise our awareness for that, and in particular make an effort to be better when it comes to new code that we write. For some parts of the library, such micro-optimization might not be feasible at the moment because the code is likely to change a lot anyway (such as everything regarding PSA Crypto, or the messaging layer), but for parts such as ASN.1 parsing or X.509, I think it's worth making improvements as we see them, and to pay attention when adding new code.

A slightly shorter way of writing this function would e.g. be the following, which saves around 1/3 of code in my system (I'm sure you'd be able to make it even shorter if you tried):
```c
int mbedtls_asn1_write_named_bitstring( unsigned char **p,
                                        unsigned char *start,
                                        const unsigned char *buf,
                                        size_t bits )
{
    if( bits != 0 )
    {
        size_t const byte_len = ( bits + 7 ) / 8;
        size_t const unused_bits = ( byte_len * 8 ) - bits;

        unsigned char cur_byte_shifted;
        const unsigned char *cur_byte;

        /*
         * Named bitstrings require that trailing 0s are excluded in the encoding
         * of the bitstring. Trailing 0s are considered part of the 'unused' bits
         * when encoding this value in the first content octet
         */
        cur_byte         = buf + byte_len - 1;
        cur_byte_shifted = *cur_byte >> unused_bits;

        do
        {
            if( ( cur_byte_shifted & 0x1 ) != 0 )
                break;

            bits--;

            cur_byte_shifted >>= 1;
            if( bits % 8 != 0 )
                cur_byte_shifted = *--cur_byte;

        } while( bits > 0 );
    }
    return( mbedtls_asn1_write_bitstring( p, start, buf, bits ) );
}
```

But again, this is just a suggestion.